File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:

    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return;
        def reverse(node):
            prev = None
            curr = node;
            while curr:
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp
            return prev
        
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next;
            fast = fast.next.next;

        second = slow.next
        slow.next = None

        node = reverse(second)
        temp = head;

        while node and temp:
            t1 = temp.next
            t2 = node.next
            temp.next = node
            node.next = t1
            temp = t1
            node = t2

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        let prev = null;
        let curr = head;
        let currNext;

        while(curr){
            currNext = curr.next;
            curr.next = prev;
            prev = curr;
            curr = currNext;
        }

        return prev;
    }
}
